pkg/registry/core/pod/strategy.go

```diff
@@ -717,3 +720,12 @@ func applySeccompVersionSkew(pod *api.Pod) {
 			return true
 		})
 }
+
+// setDefaultHostname sets the pod's hostname field if needed.
+func setDefaultHostname(pod *api.Pod) {
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.PodHostnameWhenSubdomain) {
+		if pod.Spec.Subdomain != "" && pod.Spec.Hostname == "" {
+			pod.Spec.Hostname = pod.Name
```

Comment:
ref https://github.com/kubernetes/kubernetes/issues/104195#issuecomment-908282093 :/

Reply:
so a pod.Name can be a valid subdomain

https://github.com/kubernetes/kubernetes/blob/edb0a72cff0e43bab72a02cada8486d562ee1cd5/pkg/apis/core/validation/validation.go#L225-L228

but pod.Spec.Hostname validation is restricted to a DNS label ,

https://github.com/kubernetes/kubernetes/blob/edb0a72cff0e43bab72a02cada8486d562ee1cd5/pkg/apis/core/validation/validation.go#L3451-L3453

label is a subset of subdomain IIUIC

https://github.com/kubernetes/kubernetes/blob/c1153d3353bd4f4b68d85245d53d2745586be474/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L178-L209

Reply:
however, we should put some effort to unravel this complex DNS knot , the use cases on the issues looks legit,

Reply:
Commented on #104195, new commit added here to PoC